lib/fizzy/execute.cpp

```diff
@@ -359,17 +359,17 @@ inline bool store_into_memory(
 template <typename Op>
 inline void unary_op(OperandStack& stack, Op op) noexcept
 {
-    using T = decltype(op(stack.top()));
-    stack.top() = op(static_cast<T>(stack.top()));
+    using T = decltype(op({}));
+    stack.top() = op(stack.top().as<T>());
 }
 
 template <typename Op>
 inline void binary_op(OperandStack& stack, Op op) noexcept
 {
-    using T = decltype(op(stack.top(), stack.top()));
-    const auto val2 = static_cast<T>(stack.pop());
-    const auto val1 = static_cast<T>(stack.top());
-    stack.top() = static_cast<std::make_unsigned_t<T>>(op(val1, val2));
+    using T = decltype(op({}, {}));
```

Comment:
Hm, what does this do with `{}, {}`?

Reply:
The `{}`  in `f(T)` more/less mean "construct the required type T with default constructor".

E.g. The `void func(void*, Options, vector<int>, int)` can be invoked as `func(nullptr, {}, {}, 0)` or even `func({}, {}, {}, {})`.

Reply:
Yeah I get that, but in the case of `op`, which is `std::add`, `std::sub`, etc.

Reply:
Well, `op` is an instance of e.g. `std::add<uint64_t>{}` which has `operator()` overloaded so `op(...)` expression is valid. The `decltype` gets the type of this expression.

Reply:
Okay so this line change is not related to the current PR. Could have been like this prior.

Reply:
Yes. The `{}` could have been use instead of `stack.top()` before. But now this simplifications is required as `stack.top()` has wrong type in general.